Dynamically generate format lookup table from drm_fourcc header

soreau opened the pull request:
This eliminates the need to maintain a list of formats used
for converting from fourcc codes to the prettier version by
parsing drm_fourcc.h directly and building a header for use
in pretty.c

emersion commented on fourcc_header_gen.c:
```diff
@@ -0,0 +1,164 @@
+static struct format *formats;
+static unsigned int format_count;
+
+void get_fourcc(unsigned int j, size_t size, char *map)
```
Why do we need to parse the fourcc code? We only want to generate this kind of code:

```c
switch (format) {
case DRM_FORMAT_XXX:
    return "XXX";
…
}
```

soreau replied:
You're right, this didn't occur to me until you brought it out.

emersion commented:
If we auto-generate code for formats, we should probably do the same for modifiers too.

emersion commented:
For reference, here's a similar Python script that generates Go: https://git.sr.ht/~emersion/go-drm/tree/master/fourcc.py

Here we don't need the `#define` values so it's not necessary to go as far as compiling and running a C executable to get them. Also the `pkg-config` stuff can be taken from Meson (as you do currently).

soreau commented:
> For reference, here's a similar Python script that generates Go: https://git.sr.ht/~emersion/go-drm/tree/master/fourcc.py

This was a huge help, thanks.

> If we auto-generate code for formats, we should probably do the same for modifiers too.

Done.

> I'd accept something which is written in Python 3

Done.

emersion commented on fourcc.py:
```diff
@@ -0,0 +1,51 @@
+			continue
+		ident_list.append(ident)
+
+c_file_name = "tables.h"
```
This should probably be read from `argv[1]`

soreau replied:
I was thinking the same but not sure if it was worth the effort.

emersion commented on fourcc.py:
```diff
@@ -0,0 +1,51 @@
+import subprocess
+import sys
+
+cc = os.getenv("CC", "cc")
```
This isn't used

soreau replied:
Ok.

emersion commented on fourcc.py:
```diff
@@ -0,0 +1,51 @@
+cc = os.getenv("CC", "cc")
+pkg_config = os.getenv("PKG_CONFIG", "pkg-config")
+
+cflags = subprocess.check_output([pkg_config, "libdrm", "--cflags-only-I"]).decode().strip().split()
```
Getting the directory from Meson and passing it as a parameter to the script is probably better

soreau replied:
Got it.

ascent12 commented on fourcc.py:
```diff
@@ -0,0 +1,42 @@
+ident_list = []
+with open(fourcc_include) as f:
+	for l in f.readlines():
+		parts = l.split()
```
I'd still prefer something regex-based instead of ad-hoc parsing.
```python
fmt = re.compile(r'^#define (\w+)\s*(?:\\$\s*)?fourcc_code', flags=re.M)
mod = re.compile(r'^#define (\w+)\s*(?:\\$\s*)?fourcc_mod_code', flags=re.M)
```
`DRM_FORMAT_INVALID` would be a special case, as would some of the non-trivial modifiers.

ascent12 commented on fourcc.py:
```diff
@@ -0,0 +1,42 @@
+f_name = sys.argv[2]
+
+with open(f_name, "w+") as f:
+	f.write('#include <stdint.h>\n')
```
Use multi-line strings instead of just writing out a bunch of separate write calls.
```python
'''\
#include <stdint.h>
#include <drm_fourcc.h>
'''
```
It makes the C code a lot cleaner and easier to read.

ascent12 commented on fourcc.py:
```diff
@@ -0,0 +1,42 @@
+
+f_name = sys.argv[2]
+
+with open(f_name, "w+") as f:
```
This just needs `w` instead of `w+`.
Also, can you consistently use the same type of quotes for all strings instead of mixing `'` and `"`.

ascent12 commented on fourcc.py:
```diff
@@ -0,0 +1,42 @@
+	f.write('#include <stdint.h>\n')
+	f.write('#include <drm_fourcc.h>\n')
+	f.write('\n')
+	f.write('const char *format_str(uint32_t format)\n{\n')
```
If a function is going to exist in a header, make it `static inline`.

ascent12 commented on meson.build:
```diff
@@ -23,8 +23,15 @@ config_h = configure_file(
 )
 
+tables_h_name = 'tables.h'
+drm_fourcc_header_path = libdrm.get_pkgconfig_variable('includedir') + '/libdrm/drm_fourcc.h'
```
Use `join_path`

ascent12 commented on meson.build:
```diff
@@ -23,8 +23,15 @@ config_h = configure_file(
   output: 'config.h',
 )
 
+tables_h_name = 'tables.h'
```
This doesn't need to exist. Just write it directly in the `custom_target`.

ascent12 commented on meson.build:
```diff
@@ -23,8 +23,15 @@ config_h = configure_file(
+
+tables_h = custom_target('tables_h',
+  output : tables_h_name,
+  command : [meson.source_root() + '/fourcc.py', drm_fourcc_header_path, tables_h_name, '@OUTPUT@'])
```
Use the meson python module to find the python installation, and call this using it.

ascent12 commented on fourcc.py:
```diff
@@ -0,0 +1,42 @@
+	f.write('\t\tdefault:\n\t\t\treturn \"unknown\";\n')
+	f.write('\t}\n}\n\n')
+	f.write('const char *modifier_str(uint64_t modifier)\n{\n')
+	f.write('\t/*\n\t * ARM has a complex format which we can\'t be bothered to parse.\n\t */\n')
```
Now that this is happening at build time, we could parse/generate all of the combinations for formats like this.
Although it doesn't have to be done as part of this PR; I'm just thinking out loud.

soreau commented:
@ascent12 I think I've addressed all your concerns, let me know if there's anything else.

The build bot doesn't seem to be able to import python module but it builds fine here.

ascent12 commented on meson.build:
```diff
@@ -23,8 +23,16 @@ config_h = configure_file(
 )
 
+drm_fourcc_header_path = join_paths(libdrm.get_pkgconfig_variable('includedir'), 'libdrm/drm_fourcc.h')
+pymod = import('python')
```
>The build bot doesn't seem to be able to import python module but it builds fine here.

It's so frustrating how old some debian packages can be. Can you change this to the (now deprecated) `python3` module instead?

We'll change back it when the new debian stable is released.

soreau replied:
I did but it didn't seem to help.

ascent12 replied:
God damn it debian...
They have 0.37 in their repos, and the python3 module was added in 0.38.
I suppose it'll just have to be changed back to how this originally was, without the meson modules.

soreau replied:
Done

ascent12 commented on meson.build:
```diff
@@ -23,8 +23,16 @@ config_h = configure_file(
+
+tables_h = custom_target('tables_h',
+  output : 'tables.h',
+  command : [py3, join_paths(meson.source_root(), 'fourcc.py'), drm_fourcc_header_path, 'tables.h', '@OUTPUT@'])
```
You don't need the `'tables.h'` part here. `'@OUTPUT@'` will handle that.

soreau replied:
If I remove `output : 'tables.h',` line I get `ERROR: Missing keyword argument "output".`

ascent12 replied:
No, you keep the `output: 'tables.h'` argument. Just remove it from the command array.
`@OUTPUT@` will automatically be replaced by it, and allows you to not repeat yourself in the definition.

soreau replied:
I see, that works

ascent12 commented on fourcc.py:
```diff
@@ -0,0 +1,67 @@
+	for l in f.readlines():
+		ident = fmt.search(l)
+		if ident != None:
+			add_ident(ident.group(0), fmt_list)
```
The subgroup capture I added in that regex was so this wasn't needed. Just use `.group(1)` (or `[1]` as a shortcut).

soreau replied:
I used group(1) because the build bot didn't like [1]

ascent12 commented on fourcc.py:
```diff
@@ -0,0 +1,67 @@
+			return "INVALID";
+''')
+	for ident in fmt_list:
+		f.write('\t\tcase ' + ident + ':\n\t\t\treturn "' + ident[11:] + '";\n')
```
I think this would look a lot cleaner as a format string.
```python
f.write('\t\tcase {}: return "{}";\n'.format(ident, ident[11:])
```
Also, I'm not a huge fan of the seemingly magic number 11. `len('DRM_FORMAT_')` is more obvious.

emersion commented on fourcc.py:
```diff
@@ -0,0 +1,62 @@
+mod_list = []
+
+fmt = re.compile(r'^#define (\w+)\s*(?:\\$\s*)?fourcc_code', flags=re.M)
+mod = re.compile(r'^#define (\w+)\s*(?:\\$\s*)?fourcc_mod_code', flags=re.M)
```
This missing some modifiers like `DRM_FORMAT_MOD_BROADCOM_SAND256`.

```c
#define DRM_FORMAT_MOD_BROADCOM_SAND256 \
    DRM_FORMAT_MOD_BROADCOM_SAND256_COL_HEIGHT(0)
```

ascent12 replied:
Those should probably be added back as a special case.
```python
mod_broadcom = re.compile(r'^#define (\w+)\s*(?:\\$\s*)?DRM_FORMAT_MOD_BROADCOM', flags=re.M)

emersion replied:
Why not match using the `DRM_FORMAT_MOD_` prefix (and add a special case for i915 which has a different prefix)?

ascent12 replied:
I'm trying to be somewhat defensive about what modifiers will be added in the future. We already have things like `AFBC_FORMAT_MOD_*` which are not modifiers themselves, but flags to another macro, so matching `_FORMAT_MOD_` isn't reliable. I'm just waiting for
```c
#define DRM_FORMAT_MOD_VENDOR_FLAG (1ULL)
```
to happen.

I think matching against `fourcc_mod_code` is a more reliable way to truly match things that are modifiers, but that doesn't match against macros that are defined in terms of other macros.

soreau replied:
> `mod_broadcom = re.compile(r'^#define (\w+)\s*(?:\\$\s*)?DRM_FORMAT_MOD_BROADCOM', flags=re.M)`

This doesn't seem to match anything.

soreau replied:
> This missing some modifiers like `DRM_FORMAT_MOD_BROADCOM_SAND256`.

I was able to get results using
`mod_broadcom = re.compile(r'^#define (DRM_FORMAT_MOD_BROADCOM\w+[^\w(])', flags=re.M)`

@emersion Can you see if the latest iteration does The Right Thing?

emersion replied:
The latest iteration's output LGTM. It does add some formats we were missing.

emersion commented on meson.build:
```diff
@@ -23,8 +23,14 @@ config_h = configure_file(
+
 executable('drm_info',
-  files('main.c', 'json.c', 'pretty.c'),
+  [files('main.c', 'json.c', 'pretty.c'), tables_h],
```
This nested syntax looks weird to me. Regardless, I don't think we actually need to add the header here. Meson will pick it up because it's `#include`'ed in the C file.

soreau replied:
It is required here. If I do not put it here and attempt to build after removing the build directory, I get `../pretty.c:13:10: fatal error: tables.h: No such file or directory`.

emersion replied:
Hmm, I see. It doesn't seem like you can put it in the `files()` list, can you?

soreau replied:
No, it must be the custom_target. Otherwise the header is simply not generated.

soreau commented:
@emersion Thanks for the review.

I found that `'^#define (DRM_FORMAT_MOD_BROADCOM\w+(?=\s))'` is more correct and doesn't require stripping the result.

ascent12 commented:
Thanks
